File: Week_06/minPathSum.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m=len(grid)
        n=len(grid[0])
        dp=grid
        for i in range(m):
            for j in range(n):
                if i==0 and j==0:
                    dp[i][j]=grid[i][j]
                elif i==0 and j!=0:
                    dp[i][j]=grid[i][j]+dp[i][j-1]
                elif i!=0 and j==0:
                    dp[i][j]=grid[i][j]+dp[i-1][j]
                else:
                    dp[i][j]=grid[i][j]+min(dp[i][j-1],dp[i-1][j])
        return dp[m-1][n-1]

File: Week_06/minPathSum.go
func minPathSum(grid [][]int) int {
    var min func(a int , b int) int
    min = func(a int, b int) int{
        if a<b{
            return a
        }else{
            return b
        }
    }
    dp:=grid
    m:=len(grid)
    n:=len(grid[0])
    for i:=0;i<m;i++{
        for j:=0;j<n;j++{
            if i==0 && j==0{
                continue
            }else if i==0 && j!=0{
                dp[i][j]=dp[i][j]+dp[i][j-1]
            }else if i!=0 && j==0{
                dp[i][j]=dp[i][j]+dp[i-1][j]
            }else{
                dp[i][j]=min(dp[i-1][j],dp[i][j-1])+dp[i][j]
            }
        }
    }
    return dp[m-1][n-1]
}
